### ISSUE TITLE
Add support for external storage(Sd Card, USB Storage etc) in flutter.

### ISSUE THREAD

**Issue by badaween-ahmed** [flutter/flutter#40504]
I didn't find any method to locate external storage like Sd Card, USB Storage etc which is supported in native android. When I call this method `getExternalStorageDirectory()` which returns `/storage/emulated/0` that is Internal storage (not external). Also, add support for "Scoped Storage access". The official Android doc on [storage](https://developer.android.com/guide/topics/data/data-storage) is very well written hope to see some improvement on flutter too.

**Comment by moeiscool** [comment]
i second this

**Comment by Nailik** [comment]
Jeah very important to enabled scoped storage on flutter. I just had the issue that i updated the sdk version and Directory listSync did only return folders, no more files - without any error.

**Comment by nicoechaniz** [comment]
How is it possible that SD card support is not there from day 1...
We have been developing a p2p file sharing project in Flutter only to find out mid-way about this lack of support.
Please do implement this feature. I would be more inclined to call this a severe bug than a feature. YMMV.

**Comment by moeiscool** [comment]
> How is it possible that SD card support is not there from day 1...
> We have been developing a p2p file sharing project in Flutter only to find out mid-way about this lack of support.
> Please do implement this feature. I would be more inclined to call this a severe bug than a feature. YMMV.

i agree with you there. this needs to be added. If google thinks its a good idea to use dart then they should start modernizing it

**Comment by Daibaku9999** [comment]
Any update on this?

**Comment by moeiscool** [comment]
i’m really hoping this will launch soon, i want to release some apps that rely on this function

**Comment by EArminjon** [comment]
Yea me too... My app will be blocked without this feature...

**Comment by netwrkx** [comment]
Any progress regarding this issue

**Comment by red-ibanez** [comment]
+1

I've been trying different things, different packages and I couldn't figure it out why I didn't have access to the external SD card.  Then found out this open issue, dah!

**Comment by RisingSunStudios** [comment]
Still nothing - found out today that getExternalStorageDirectory() is returning internal storage path..

**Comment by hBurt** [comment]
So here is what i did that works for me:

use the [path_provider plugin](https://pub.dev/packages/path_provider)

here is a function I made to grab the external sd cards path on android:

`Future<Directory> getExternalSdCardPath() async {
    List<Directory> extDirectories = await getExternalStorageDirectories();

    List<String> dirs = extDirectories[1].toString().split('/');
    String rebuiltPath = '/' + dirs[1] + '/' + dirs[2] + '/';

    print("rebuilt path: " + rebuiltPath);
    return new Directory(rebuiltPath);
  }`

That function return the first item in this list: 
• 5FDD-C1A1
• emulated
• enc_emulated
• self

If you can't see it, "5FDD-C1A1" is the name to the sd cards path. SO the function above is returning: /storage/5FDD-C1A1/

with this you should be able to access you external sd card storage, this is assuming you also allow permissions in your app.
[I use this permission_handler](https://pub.dev/packages/permission_handler)

Just call this method in your initState function on your apps main page:

`void requestStoragePermissions() async {
    // You can request multiple permissions at once.
    Map<Permission, PermissionStatus> statuses = await [
      // Permission.location,
      Permission.storage,
    ].request();

    print(statuses[Permission.storage]);
  }`


Cool, so now all of that i setup, now just grab the list of all items in that dir:

`Future<List> readExternalDirectories() async {
    List files = new List();
    Directory directory = await getExternalSdCardPath();

    if (await Permission.storage.request().isGranted) {
      files = directory.listSync();
    }

    return files;
  }`

You should be good to go now!

**Comment by ZhFahim** [comment]
Any progress? I really need the feature to write file to SD card.

**Comment by moda20** [comment]
Can we write files to SDCard after getting the permissions? I built a plugin that gives my app the permissions for SDcard modification, but I still can't save files on SDcard.

**Comment by badaween-ahmed** [comment]
I think this issue should close now, lack of interest, and also the new API moving towards iOS style storage access. 😑
https://developer.android.com/about/versions/11/privacy/storage

**Comment by moda20** [comment]
Yeah, I can see that, but compatibility is needed. I still have an android 7 phone with no manufacturer support or custom ROMs.

**Comment by jeremielodi** [comment]
@hBurt I have tried your soluton but I still have the permission issue.
```
I/flutter (21085): rebuilt path: /storage/C6FC-0030/
I/flutter (21085): /storage/C6FC-0030/System Volume Information
I/flutter (21085): /storage/C6FC-0030/LOST.DIR
I/flutter (21085): /storage/C6FC-0030/.android_secure
I/flutter (21085): /storage/C6FC-0030/Android
I/flutter (21085): /storage/C6FC-0030/DCIM
I/flutter (21085): /storage/C6FC-0030/backup
I/flutter (21085): /storage/C6FC-0030/id_rsa.pub
I/flutter (21085): /storage/C6FC-0030/music
I/flutter (21085): /storage/C6FC-0030/images
I/flutter (21085): /storage/C6FC-0030/doc
I/flutter (21085): /storage/C6FC-0030/collect-settings.png
I/flutter (21085): /storage/C6FC-0030/c
I/flutter (21085): /storage/C6FC-0030/ff.apk
I/flutter (21085): /storage/C6FC-0030/video.apk
I/flutter (21085): /storage/C6FC-0030/PhoneRecord
I/flutter (21085): /storage/C6FC-0030/bluetooth
I/flutter (21085): /storage/C6FC-0030/card
I/flutter (21085): /storage/C6FC-0030/Pictures
I/flutter (21085): /storage/C6FC-0030/printer
I/flutter (21085): /storage/C6FC-0030/congoerpPharmacie.sql
I/flutter (21085): /storage/C6FC-0030/msql_queries.txt
I/flutter (21085): /storage/C6FC-0030/log.txt
I/flutter (21085): /storage/C6FC-0030/cpp
I/flutter (21085): /storage/C6FC-0030/000-default.conf
I/flutter (21085): /storage/C6FC-0030/settup.txt
I/flutter (21085): /storage/C6FC-0030/install
I/flutter (21085): /storage/C6FC-0030/ifr
I/flutter (21085): FileSystemException: Creation failed, path = '/storage/C6FC-0030/flutter' (OS Error: Permission denied, errno = 13)
```

**Comment by jeremielodi** [comment]
This is a very important bug to fixe so I wonder if everyone is aware that

**Comment by Nailik** [comment]
> This is a very important bug to fixe so I wonder if everyone is aware that

It's one of the main reasons that could speak against using flutter for an Android App. And looking at "P4" they don't seem to be aware that it's really important.
Until this is fixed flutter doesn't even really support the most basic Android Features.

**Comment by JAICHANGPARK** [comment]
@jeremielodi Are you solved this permission issue? 

> @hBurt I have tried your soluton but I still have the permission issue.
> 
> ```
> I/flutter (21085): rebuilt path: /storage/C6FC-0030/
> I/flutter (21085): /storage/C6FC-0030/System Volume Information
> I/flutter (21085): /storage/C6FC-0030/LOST.DIR
> I/flutter (21085): /storage/C6FC-0030/.android_secure
> I/flutter (21085): /storage/C6FC-0030/Android
> I/flutter (21085): /storage/C6FC-0030/DCIM
> I/flutter (21085): /storage/C6FC-0030/backup
> I/flutter (21085): /storage/C6FC-0030/id_rsa.pub
> I/flutter (21085): /storage/C6FC-0030/music
> I/flutter (21085): /storage/C6FC-0030/images
> I/flutter (21085): /storage/C6FC-0030/doc
> I/flutter (21085): /storage/C6FC-0030/collect-settings.png
> I/flutter (21085): /storage/C6FC-0030/c
> I/flutter (21085): /storage/C6FC-0030/ff.apk
> I/flutter (21085): /storage/C6FC-0030/video.apk
> I/flutter (21085): /storage/C6FC-0030/PhoneRecord
> I/flutter (21085): /storage/C6FC-0030/bluetooth
> I/flutter (21085): /storage/C6FC-0030/card
> I/flutter (21085): /storage/C6FC-0030/Pictures
> I/flutter (21085): /storage/C6FC-0030/printer
> I/flutter (21085): /storage/C6FC-0030/congoerpPharmacie.sql
> I/flutter (21085): /storage/C6FC-0030/msql_queries.txt
> I/flutter (21085): /storage/C6FC-0030/log.txt
> I/flutter (21085): /storage/C6FC-0030/cpp
> I/flutter (21085): /storage/C6FC-0030/000-default.conf
> I/flutter (21085): /storage/C6FC-0030/settup.txt
> I/flutter (21085): /storage/C6FC-0030/install
> I/flutter (21085): /storage/C6FC-0030/ifr
> I/flutter (21085): FileSystemException: Creation failed, path = '/storage/C6FC-0030/flutter' (OS Error: Permission denied, errno = 13)
> ```

**Comment by JAICHANGPARK** [comment]
Hi folks~!
Any have update this issues? 

I tried to use usbmanager 
but still not works for me

**Comment by UnicornsOnLSD** [comment]
I think this is just Android being weird and Flutter not properly supporting Android storage URIs. For now, using the external storage's app directory (which can be done in 0.4.3's custom storage manager) is the safest option.

**Comment by fiorcode** [comment]
It would be good if they solve this requirement.

**Comment by pavel-birdy** [comment]
Can you guys try this package and see if it's working or not? I gave up making such app. 
https://pub.dev/packages/file_manager

**Comment by erlangparasu** [comment]
any update?

**Comment by AyushDobariya** [comment]
List<FileSystemEntity> pdfList = [];
  get localpath async {
    Directory? dirList = await getExternalStorageDirectory();
    String pdfFile =
        "${dirList?.path.replaceAll('/files', "/files/documents")}";
    print("pdfFile :: ${pdfFile}");
    files = dirList?.listSync(recursive: true, followLinks: false);
    if (files != null) {
      for (FileSystemEntity entity in files!) {
        String path = entity.path;
        if (entity.path.endsWith(".pdf")) pdfList.add(entity);
        print("entity : ${entity.path}");
      }
    } else {
      return pdfList;
    }
    print("pdf List${pdfList}");
    print(pdfList.length);
    return dirList?.path;
  }

**Comment by Doppelklick** [comment]
This has been a problem for a long time - we can expect a solution?

**Comment by ENRexGeorge23** [comment]
It's now 2024 is anyone in here now has the solution for this ?

**Comment by ENRexGeorge23** [comment]
We finally got it, this package will make you save files in your sd card, we used it to save our local database in our application

https://pub.dev/packages/path_provider_ex2

**Comment by camelCase4** [comment]
> We finally got it, this package will make you save files in your sd card, we used it to save our local database in our application
> 
> https://pub.dev/packages/path_provider_ex2

Thanks you are a life saver! T_T
